IPython/core/history.py

```diff
@@ -81,18 +80,29 @@ def catch_corrupt_db(f, self, *a, **kw):
     try:
         return f(self, *a, **kw)
     except (DatabaseError, OperationalError) as e:
-        if os.path.isfile(self.hist_file):
-            # Try to move the file out of the way
-            base,ext = os.path.splitext(self.hist_file)
-            newpath = base + '-corrupt' + ext
-            os.rename(self.hist_file, newpath)
-            print("ERROR! History file wasn't a valid SQLite database (%s)." % e,
-            "It was moved to %s" % newpath, "and a new file created.", file=sys.stderr)
+        self._corrupt_db_counter += 1
+        self.log.error("Failed to open SQLite history %s (%s).", self.hist_file, e)
+        if self.hist_file != ':memory:':
+            if self._corrupt_db_counter > self._corrupt_db_limit:
+                self.hist_file = ':memory:'
+                self.log.error("Failed to load history too many times, history will not be saved.")
+            elif os.path.isfile(self.hist_file):
+                # Try to move the file out of the way
+                base,ext = os.path.splitext(self.hist_file)
+                now = datetime.datetime.now().isoformat().replace(':', '.')
+                newpath = base + '-corrupt-' + now + ext
```

Comment:
Right, but in the NFS case prompting this, it's failing to even create the database. So there's never a useful 'good' database worth preserving.

I'm not too concerned if starting IPython twice with a corrupt database wipes out your history - if we can't use that database, your history is effectively gone anyway, and I think ~no-one is going to take the time to recover it. I'd be happy if it kept it for a bit longer, or was smart enough not to move a database that it had just tried and failed to create. I'm just not happy about the ease with which this could create a great many useless files.